subsys/storage/flash_buffered_write/fbw.c

```diff
@@ -0,0 +1,181 @@
+
+	if (rc) {
+		LOG_ERR("Error %d while erasing page", rc);
+	}
+
+	return rc;
+}
+
+#endif /* CONFIG_FBW_ERASE */
+
+static int flash_sync(struct fbw_ctx *ctx)
+{
+	int rc = 0;
+	size_t write_addr = ctx->offset + ctx->bytes_written;
+
+
+#ifdef CONFIG_FBW_ERASE
+	rc = fbw_erase(ctx, write_addr + ctx->buf_bytes);
+	if (rc) {
+		LOG_ERR("fbw_erase error %d offset=0x%08zx", rc,
+			write_addr);
+		return rc;
+	}
+#endif
+
+	flash_write_protection_set(ctx->fdev, false);
+	rc = flash_write(ctx->fdev, write_addr, ctx->buf, ctx->buf_bytes);
+	flash_write_protection_set(ctx->fdev, true);
+
+	if (rc) {
+		LOG_ERR("flash_write error %d offset=0x%08zx", rc,
+			write_addr);
+		return rc;
+	}
+
+	if (ctx->callback) {
+		rc = flash_read(ctx->fdev, write_addr, ctx->buf,
+				ctx->buf_bytes);
+		if (rc != 0) {
+			LOG_ERR("flash read failed: %d", rc);
+			return rc;
+		}
+
+		rc = ctx->callback(ctx->buf, ctx->buf_bytes, write_addr);
+		if (rc != 0) {
+			LOG_ERR("callback failed: %d", rc);
+		}
+	}
+
+	ctx->bytes_written += ctx->buf_bytes;
+	ctx->buf_bytes = 0U;
+
+	return rc;
+}
+
+int fbw_write(struct fbw_ctx *ctx, u8_t *data, size_t len, bool flush)
+{
+	int processed = 0;
+	int rc = 0;
+	int buf_empty_bytes;
+
+	if (!ctx || !data) {
+		return -EFAULT;
+	}
+
+	if (ctx->bytes_written + ctx->buf_bytes + len > ctx->available) {
+		return -ENOMEM;
+	}
+
+	while ((len - processed) >=
+	       (buf_empty_bytes = ctx->buf_len - ctx->buf_bytes)) {
+		memcpy(ctx->buf + ctx->buf_bytes, data + processed,
+		       buf_empty_bytes);
+
+		ctx->buf_bytes = ctx->buf_len;
+		rc = flash_sync(ctx);
+
+		if (rc) {
+			return rc;
+		}
+
+		processed += buf_empty_bytes;
+	}
+
+	/* place rest of the data into ctx->buf */
+	if (processed < len) {
+		memcpy(ctx->buf + ctx->buf_bytes,
+		       data + processed, len - processed);
+		ctx->buf_bytes += len - processed;
+	}
+
+	if (flush && ctx->buf_bytes > 0) {
+		rc = flash_sync(ctx);
+	}
+
+	return rc;
+}
+
+size_t fbw_bytes_written(struct fbw_ctx *ctx)
+{
+	return ctx->bytes_written;
+}
+
+int fbw_init(struct fbw_ctx *ctx, struct device *fdev, u8_t *buf,
+	     size_t buf_len, size_t offset, size_t size, fbw_callback_t cb)
+{
+	if (!ctx || !fdev || !buf) {
+		return -EFAULT;
+	}
+
+	size_t layout_size = 0;
+	size_t total_size = 0;
+	const struct flash_pages_layout *layout;
+	const struct flash_driver_api *api = fdev->driver_api;
+
+	/* Calculate the total size of the flash device */
+	api->page_layout(fdev, &layout, &layout_size);
+	for (int i = 0; i < layout_size; i++) {
+		total_size += layout->pages_count * layout->pages_size;
+	}
+
+	if ((offset + size) > total_size ||
+	    offset % api->write_block_size) {
+		LOG_ERR("Incorrect parameter");
+		return -EFAULT;
+	}
+
+	ctx->fdev = fdev;
+	ctx->buf = buf;
+	ctx->buf_len = buf_len;
+	ctx->bytes_written = 0;
+	ctx->buf_bytes = 0U;
+	ctx->offset = offset;
+	ctx->available = (size == 0 ? total_size - offset : size);
+	ctx->callback = cb;
+#ifdef CONFIG_FBW_ERASE
```

Comment:
You can use ` if (IS_ENABLED(CONFIG_FBW_ERASE)) {`

Reply:
This actually fails with `gcc version 9.2.1 20200130 (Arch Linux 9.2.1+20200130-2)` so for now I only use `IS_ENABLED` earlier in this file. We might want to consider to only use `ifdef` to be more consistent?

Reply:
Updated. I removed the `#ifdef` surrounding the definition of the fbw_erase funtion as well.

Reply:
I'm not able to get `IS_ENABLED` to work the same way as `#ifdef`. See the error below.

```
fbw.c: In function 'fbw_init':
fbw.c:179:6: error: 'struct fbw_ctx' has no member named 'last_erased_page_start_offset'
   ctx->last_erased_page_start_offset = -1;
      ^~
```

From the lines 
```
	if (IS_ENABLED(CONFIG_FBW_ERASE)) {
		ctx->last_erased_page_start_offset = -1;
	}
```

Where as using `#ifdef CONFIG_FBW_ERASE` does not give an error.

Reply:
I guess this is because the definition of the struct is in a different file, and has the `#ifdef` surrounding the inclusion of this member variable.

Reply:
sad. Sorry I missed it.